src/Scout/ScoutEngine.php

```diff
@@ -0,0 +1,535 @@
+     *
+     * @param Builder      $builder
+     * @param array|Cursor $results
+     * @param Model        $model
+     *
+     * @return LazyCollection
+     */
+    public function lazyMap(Builder $builder, $results, $model): LazyCollection
+    {
+        assert($results instanceof Cursor || is_array($results), new TypeError(sprintf('Argument #2 ($results) must be of type %s|array, %s given', Cursor::class, get_debug_type($results))));
+        assert($model instanceof Model, new TypeError(sprintf('Argument #3 ($model) must be of type %s, %s given', Model::class, get_debug_type($model))));
+
+        if (! $results) {
+            return LazyCollection::make($model->newCollection());
+        }
+
+        $objectIds = array_column($results, '_id');
+        $objectIdPositions = array_flip($objectIds);
+
+        return $model->queryScoutModelsByIds(
+            $builder,
+            $objectIds,
+        )->cursor()->filter(function ($model) use ($objectIds) {
+            return in_array($model->getScoutKey(), $objectIds);
+        })->map(function ($model) use ($results, $objectIdPositions) {
+            $result = $results[$objectIdPositions[$model->getScoutKey()]] ?? [];
+
+            foreach ($result as $key => $value) {
+                if (substr($key, 0, 1) === '_') {
+                    $model->withScoutMetadata($key, $value);
+                }
+            }
+
+            return $model;
+        })->sortBy(function ($model) use ($objectIdPositions) {
+            return $objectIdPositions[$model->getScoutKey()];
+        })->values();
+    }
+
+    /**
+     * Create the MongoDB Atlas Search index.
+     *
+     * Accepted options:
+     *  - wait: bool, default true. Wait for the index to be created.
+     *
+     * @param string            $name    Collection name
+     * @param array{wait?:bool} $options
+     */
+    public function createIndex($name, array $options = []): void
+    {
+        assert(is_string($name), new TypeError(sprintf('Argument #1 ($name) must be of type string, %s given', get_debug_type($name))));
+
+        // Ensure the collection exists before creating the search index
+        $this->mongodb->createCollection($name);
+        $collection = $this->mongodb->selectCollection($name);
+
+        $collection = $this->mongodb->selectCollection($name);
+        /** @todo accept configuration for the mapping */
+        $collection->createSearchIndex(
+            $options['definition'] ?? self::DEFAULT_DEFINITION,
+            ['name' => self::INDEX_NAME],
+        );
+
+        if ($options['wait'] ?? true) {
+            $this->wait(function () use ($collection) {
+                $indexes = $collection->listSearchIndexes([
+                    'name' => self::INDEX_NAME,
+                    'typeMap' => ['root' => 'bson'],
+                ]);
+
+                // Many indexes with the same name may exist on Atlas local
```

Comment:
Why would multiple indexes with the same name exist? [createSearchIndexes](https://www.mongodb.com/docs/manual/reference/command/createSearchIndexes/) says:

> You cannot create multiple indexes with the same name on a single collection.

Reply:
I had a but with atlas local previously, but it seems to be resolved.